File: iteration2/Registration.java
package iteration2;


import java.util.ArrayList;
import java.util.List;

public class Registration {
    /**
     * Represents a student.
     */
    private Student student;
    /**
     * Represents a list of courses.
     * The list is stored as a private member variable of type {@code List<Course>} and can be accessed using appropriate getter and setter methods.
     */
    private List<Course> courses;

    /**
     * Constructs a new Registration object.
     *
     * @param student the student for whom the registration is being created
     * @param courses the list of courses the student is registering for
     */
    public Registration(Student student, List<Course> courses) {
        this.student = student;
        this.courses = courses;
    }

    /**
     * Approves the request for registration.
     * This method updates the student's transcript and sets the status of the request to "approved".
     * It also clears the student's draft and prints a message confirming the approval.
     */
    public void approveRequest() {
        ConsoleColours.paintYellowMenu();
        System.out.println("Request approved.");
        ConsoleColours.resetColour();
        CourseRegistrationSystem system = new CourseRegistrationSystem();

        for (Course course : courses) {
            //Student will add a course that he has never taken before for the first time
            if (!student.getTranscript().getStudentCourses().contains(course)) {
                ArrayList<Grade> tempGrade = new ArrayList<>();
                system.addToSchedule(course, student);
                tempGrade.add(null);
                student.getTranscript().getCourseGradeMap().put(course, tempGrade);
                student.getTranscript().getStudentCourses().add(course);
                course.setNumberOfStudents(course.getNumberOfStudents() + 1);
            } else {

                List<Grade> myTempList = student.getTranscript().getCourseGradeMap().get(course);

                //If there is only one grade and it is null, student added it to the transcript with
                // 'add' a moment ago, now he will cancel and delete it
                if (myTempList.size() == 1 && myTempList.getFirst() == null) {
                    system.removeFromSchedule(course, student);
                    student.getTranscript().getCourseGradeMap().remove(course);
                    student.getTranscript().getStudentCourses().remove(course);
                    course.setNumberOfStudents(course.getNumberOfStudents() - 1);
                } else if (myTempList.size() == 1 && myTempList.getFirst() != null) {
                    system.addToSchedule(course, student);
                    myTempList.add(null);
                    course.setNumberOfStudents(course.getNumberOfStudents() + 1);
                }

                //Student had a grade before, and he had just taken the course. Now, he changed his mind,
                // saying it would be too challenging, and deleted it
                else if (myTempList.size() > 1 && myTempList.getLast() == null) {
                    system.removeFromSchedule(course, student);
                    myTempList.remove(myTempList.getLast());
                    course.setNumberOfStudents(course.getNumberOfStudents() - 1);
                }

                //He has a previous grade, and the final grade is not --, which means
                // he will decide not to take the course again
                else if (!myTempList.isEmpty() && myTempList.getLast() != null) {
                    system.addToSchedule(course, student);
                    myTempList.add(null);
                    course.setNumberOfStudents(course.getNumberOfStudents() + 1);
                }
            }
        }
        student.setHasRequest(false);
        student.getDraft().clear();
        (new Notification(student, "Your request has been approved.")).sendNotification();
    }

    /**
     * Rejects a request by a student.
     * Clears the draft and marks the request as rejected.
     */
    public void rejectRequest() {
        ConsoleColours.paintYellowMenu();
        System.out.println("Request rejected.");
        ConsoleColours.resetColour();
        System.out.println();

        student.setHasRequest(false);
        student.getDraft().clear();

        (new Notification(student, "Your request has been rejected.")).sendNotification();
    }

    /**
     * Adds a request to the advisor's list of requests and sets the student's request status to true.
     * Prints a message indicating that the request has been sent to the advisor.
     *
     * @param advisor the advisor to send the request to
     */
    public void addRequest(Advisor advisor) {
        advisor.getRequests().add(this);
        student.setHasRequest(true);
        ConsoleColours.paintYellowMenu();
        System.out.println("Request sent to advisor.");
        ConsoleColours.resetColour();
    }

    public Student getStudent() {
        return student;
    }

    public List<Course> getCourses() {
        return courses;
    }
}

File: iteration3/Registration.py
from iteration3.ConsoleColours import ConsoleColours
from iteration3.Notification import Notification


class Registration:

    def __init__(self, student, courses):
        from iteration3.CourseRegistrationSystem import CourseRegistrationSystem
        self.__system = CourseRegistrationSystem()
        self.__student = student
        self.__courses = courses

    def approveRequest(self):
        ConsoleColours.paintYellowMenu()
        print("Request approved.")
        ConsoleColours.resetColour()
        for course in self.__courses:
            if course not in self.getStudent().getTranscript().getStudentCourses():
                self.__system.addToSchedule(course, self.getStudent())
                self.getStudent().getTranscript().getCourseGradeMap()[course] = [None]
                self.getStudent().getTranscript().getStudentCourses().append(course)
                course.setNumberOfStudents(course.getNumberOfStudents() + 1)
            else:
                tempList = self.getStudent().getTranscript().getCourseGradeMap()[course]

                if len(tempList) == 1 and tempList[0] is None:
                    self.__system.removeFromSchedule(course, self.getStudent())
                    self.getStudent().getTranscript().getCourseGradeMap().pop(course, None)
                    self.getStudent().getTranscript().getStudentCourses().remove(course)
                    course.setNumberOfStudents(course.getNumberOfStudents() - 1)
                elif len(tempList) == 1 and tempList[0] is not None:
                    self.__system.addToSchedule(course, self.getStudent())
                    tempList.append(None)
                    course.setNumberOfStudents(course.getNumberOfStudents() + 1)
                elif len(tempList) > 1 and tempList[-1] is None:
                    self.__system.removeFromSchedule(course, self.getStudent())
                    tempList.remove(tempList[-1])
                    course.setNumberOfStudents(course.getNumberOfStudents() - 1)
                elif tempList and tempList[-1] is not None:
                    self.__system.addToSchedule(course, self.getStudent())
                    tempList.append(None)
                    course.setNumberOfStudents(course.getNumberOfStudents() + 1)

        self.getStudent().setHasRequest(False)
        self.getStudent().getDraft().clear()
        Notification(self.getStudent(), "Your request has been approved.").sendNotification()

    def rejectRequest(self):
        ConsoleColours.paintYellowMenu()
        print("Request rejected.")
        ConsoleColours.resetColour()
        self.getStudent().setHasRequest(False)
        self.getStudent().getDraft().clear()

        Notification(self.getStudent(), "Your request has been rejected.").sendNotification()

    def addRequest(self, advisor):
        advisor.getRequestList().append(self)
        self.getStudent().setHasRequest(True)
        ConsoleColours.paintYellowMenu()
        print("Request sent to advisor.")
        ConsoleColours.resetColour()
    def getStudent(self):
        return self.__student

    def getCourses(self):
        return self.__courses
